lib/mongo/auth/stringprep.rb

```diff
@@ -0,0 +1,69 @@
+require 'mongo/auth/stringprep/tables'
+require 'mongo/error/string_prep'
+
+module Mongo
+  module Auth
+    module StringPrep
+      extend self
+
+      def prepare(data, mappings, prohibited, options = {})
+        out = apply_maps(data, mappings)
+        normalize(out) if options[:normalize]
+        check_prohibited(out, prohibited)
+        check_bidi(out) if options[:bidi]
+        out
+      end
+
+      private
+
+      def apply_maps(data, mappings)
+        out = ''
+
+        data.each_char do |c|
+          out << mapping(c, mappings)
+        end
+
+        out
+      end
+
+      def check_bidi(out)
+        if out.any? { |c| table_contains?(Tables::C8, c) }
+          raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
```

Comment:
This should be Error::StringPrep.new(Error::StringPrep::INVALID_BIDIRECTIONAL)

Reply:
Oops! I've been away from Ruby for too long...